### PR TITLE
EDM-578: freeze all config types

### PR DESCRIPTION
**Pull request by avishayt**
This PR is split into several commits for a more straightforward review.

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -457,6 +457,8 @@ github.com/stretchr/testify v1.9.0 h1:HtqpIVDClZ4nwg75+f6Lvsy/wHu+3BoSGCbBAcpTsT
 github.com/stretchr/testify v1.9.0/go.mod h1:r2ic/lqez/lEtzL7wO/rwa5dbSLXVDPFyf8C91i36aY=
 github.com/ugorji/go/codec v1.2.11 h1:BMaWp1Bb6fHwEtbplGBGJ498wD+LKlNSl25MjdZY4dU=
 github.com/ugorji/go/codec v1.2.11/go.mod h1:UNopzCgEMSXjBc6AOMqYvWC1ktqTAfzJZUZgYf6w6lg=
+github.com/valkey-io/valkey-go v1.0.48 h1:FSkZbS8X852icAWDfsBinQe0kUGO9+p/9Qj7bgRNHTw=
+github.com/valkey-io/valkey-go v1.0.48/go.mod h1:BXlVAPIL9rFQinSFM+N32JfWzfCaUAqBpZkc4vPY6fM=
```
Comment (hexfusion) [inline go.sum:461]: This is a fairly large architectural change. Looking at EDM-578, I don't see any detail on why a k/v store is needed or why this one was chosen.

File: internal/service/templateversion.go [hunk 2]
```diff
@@ -90,6 +91,27 @@ func (h *ServiceHandler) ListTemplateVersions(ctx context.Context, request serve
 // (DELETE /api/v1/api/v1/fleets/{fleet}/templateVersions)
 func (h *ServiceHandler) DeleteTemplateVersions(ctx context.Context, request server.DeleteTemplateVersionsRequestObject) (server.DeleteTemplateVersionsResponseObject, error) {
 	orgId := store.NullOrgId
+	listParams := store.ListParams{Limit: 100}
```
Comment (hexfusion) [inline internal/service/templateversion.go:94]: can we add a comment why limit is hardcoded? If 100 is an important number maybe can we add it as a const?

File: internal/store/templateversion.go [hunk 3]
```diff
@@ -57,13 +56,6 @@ func (s *TemplateVersionStore) Create(ctx context.Context, orgId uuid.UUID, reso
 	templateVersion.OrgID = orgId
 	templateVersion.Generation = lo.ToPtr[int64](1)
 	templateVersion.ResourceVersion = lo.ToPtr[int64](1)
-	status := api.TemplateVersionStatus{}
-	api.SetStatusCondition(&status.Conditions, api.Condition{Type: api.TemplateVersionValid, Status: api.ConditionStatusUnknown})
-	templateVersion.Status = model.MakeJSONField(status)
-	fleet := model.Fleet{Resource: model.Resource{OrgID: orgId, Name: resource.Spec.Fleet}}
-	if err = s.db.First(&fleet).Error; err != nil {
-		return nil, ErrorFromGormError(err)
-	}
```
Comment (hexfusion) [inline internal/store/templateversion.go:66]: question: just high level why was this removed here?

File: internal/tasks/config_storage.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+func (k *ConfigStorageHttpKey) ComposeKey() string {
+	md5sum := md5.Sum([]byte(k.URL)) //nolint: gosec
+	return fmt.Sprintf("v1/%s/%s/%s/http/%x", k.OrgID, k.Fleet, k.TemplateVersion, md5sum)
+}
+
+type ConfigStorage interface {
+	Close()
+	StoreIfNotExistsAndFetch(ctx context.Context, key string, value []byte) ([]byte, error)
+	StoreIfNotExists(ctx context.Context, key string, value []byte) error
```
Comment (hexfusion) [inline internal/tasks/config_storage.go:76]: Could we make evaluations outside the methods and then use a simplified interface WDYT?

```go
type ConfigStorage interface {
	Close()
	Store(ctx context.Context, key string, value []byte) error
	Fetch(ctx context.Context, key string) ([]byte, error)                      
	Delete(ctx context.Context, key string) error
	DeleteByPrefix(ctx context.Context, prefix string) error 
	[..]
```

**Review by hexfusion — COMMENTED** [review]
Couple questions on first pass:
- is the k/v store considered optional if not do we need to update the docs/ ?
- if it's not optional does it need to be wrapped in a binary via cmd?
- what is the horizontal scaling story?
- does this require redis or does it work with redis?
- go-redis seems to be very mature (20k likes) (307) contributors.

File: go.sum (hunk 1 above)
Comment (avishayt) [reply go.sum:461]: Right.  Added a link to the HLD doc in JIRA:
https://docs.google.com/document/d/1f_ZkgoNUtsU55AOyB_BjAZiLSDoaWF3rmS250VZLxmI/edit?usp=sharing

File: internal/tasks/config_storage.go (hunk 4 above)
Comment (avishayt) [reply internal/tasks/config_storage.go:76]: What do you mean by "evaluations"?

Comment (hexfusion) [reply internal/tasks/config_storage.go:76]: it seems that the methods do more than one thing. where you can have a simple interface and `Fetch()  then based on evaluation of result `Store()`. Just as an example.

Comment (hexfusion) [reply internal/tasks/config_storage.go:76]: This isn't blocking just an observation

**Comment by hexfusion** [comment]
> Which binary should be wrapped?
> 
> >

My bad I thought the server was also written in golang.

File: deploy/helm/flightctl/templates/flightctl-kv-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
+apiVersion: apps/v1
+kind: Deployment
```
Comment (asafbss) [inline deploy/helm/flightctl/templates/flightctl-kv-deployment.yaml:2]: Should we use StatefulSet?

File: deploy/helm/flightctl/templates/flightctl-kv-service.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+spec:
+  {{- if and .Values.global.nodePorts.kv (eq (include "flightctl.getServiceExposeMethod" .) "nodePort") }}
+  type: NodePort
+  {{- end }}
+  ports:
+    - name: "6379"
+      port: 6379
+      targetPort: 6379
+      {{- if and .Values.global.nodePorts.kv (eq (include "flightctl.getServiceExposeMethod" .) "nodePort") }}
+      nodePort: {{ .Values.global.nodePorts.kv }}
```
Comment (asafbss) [inline deploy/helm/flightctl/templates/flightctl-kv-service.yaml:17]: ClusterIP for internal communication only?

File: internal/tasks/config_storage.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+			return nil, fmt.Errorf("failed fetching key: %w", err)
+		}
+
+		return storeVal, nil
+	}
+	return value, nil
+}
+
+func (s *configStorage) StoreIfNotExists(ctx context.Context, key string, value []byte) error {
+	return s.client.Do(ctx, s.client.B().Set().Key(key).Value(valkey.BinaryString(value)).Nx().Build()).Error()
```
Comment (asafbss) [inline internal/tasks/config_storage.go:120]: Thread safety is guaranteed?

File: internal/store/templateversion.go [hunk 8]
```diff
@@ -22,8 +22,7 @@ type TemplateVersion interface {
 	DeleteAll(ctx context.Context, orgId uuid.UUID, fleet *string) error
 	Get(ctx context.Context, orgId uuid.UUID, fleet string, name string) (*api.TemplateVersion, error)
 	Delete(ctx context.Context, orgId uuid.UUID, fleet string, name string) error
-	UpdateStatus(ctx context.Context, orgId uuid.UUID, resource *api.TemplateVersion, valid *bool, callback TemplateVersionStoreCallback) error
-	GetNewestValid(ctx context.Context, orgId uuid.UUID, fleet string) (*api.TemplateVersion, error)
+	GetNewest(ctx context.Context, orgId uuid.UUID, fleet string) (*api.TemplateVersion, error)
```
Comment (hexfusion) [inline internal/store/templateversion.go:25]: optional:
```suggestion
	GetLatest(ctx context.Context, orgId uuid.UUID, fleet string) (*api.TemplateVersion, error)
```

File: internal/tasks/config_storage.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+		end
+		return value
+	`
+	getSetNxScript := valkey.NewLuaScript(luaScript)
+
+	return &configStorage{client: client, getSetNxScript: getSetNxScript}, nil
+}
+
+func (s *configStorage) Close() {
+	s.client.Close()
```
Comment (hexfusion) [inline internal/tasks/config_storage.go:105]: are calls to Close idempotent?

File: internal/tasks/device_render.go [hunk 10]
```diff
@@ -282,134 +250,230 @@ func validateParameters(item RenderItem, validateOnly, deviceBelongsToFleet bool
...
-	for name, contents := range secret.Data {
+	for name, contents := range secretData {
 		ignitionWrapper.SetFile(filepath.Join(append(splits, name)...), contents, 0o644, false, nil, nil)
 	}
-	if !args.validateOnly {
-		args.ignitionConfig = lo.ToPtr(ignitionWrapper.Merge(*args.ignitionConfig))
-	}
-	return k8sSpec.Name, nil
+
+	*ignitionConfig = lo.ToPtr(ignitionWrapper.Merge(**ignitionConfig))
```
Comment (hexfusion) [inline internal/tasks/device_render.go:464]:
```suggestion
        configValue := *ignitionConfig
        mergedConfig := ignitionWrapper.Merge(configValue)
       *ignitionConfig = lo.ToPtr(mergedConfig)
```

Comment (hexfusion) [inline internal/tasks/device_render.go:464]: should we consider moving ToPtr to util?

File: internal/tasks/device_render.go [hunk 11]
```diff
@@ -282,134 +250,230 @@ func validateParameters(item RenderItem, validateOnly, deviceBelongsToFleet bool
...
+	return &k8sSpec.Name, nil, nil
 }
 
-func renderInlineConfig(configItem *api.ConfigProviderSpec, args *renderConfigArgs) (string, error) {
+func (t *DeviceRenderLogic) renderInlineConfig(configItem *api.ConfigProviderSpec, ignitionConfig **config_latest_types.Config) (*string, *string, error) {
```
Comment (hexfusion) [inline internal/tasks/device_render.go:468]:
```
ignitionConfig **config_latest_types.Config
```
Why does this have to be a pointer of a pointer?

File: internal/tasks/fleet_validate.go [hunk 12]
```diff
@@ -99,3 +110,140 @@ func (t *FleetValidateLogic) setStatus(ctx context.Context, validationErr error)
...
+		configItem := (*t.templateConfig)[i]
+		name, repoName, err := t.validateConfigItem(ctx, &configItem)
+		paramErr := validateParameterFormatInConfig(&configItem)
+
+		if repoName != nil {
+			referencedRepos = append(referencedRepos, *repoName)
+		}
+
+		// An error message regarding invalid parameters should take precedence
+		// because it may be the cause of the render error
```
Comment (hexfusion) [inline internal/tasks/fleet_validate.go:132]: but don't we care about the original error at all?

File: test/scripts/deploy_with_helm.sh [hunk 13]
```diff
@@ -5,6 +5,8 @@ METHOD=install
 ONLY_DB=
 RABBITMQ_VERSION=${RABBITMQ_VERSION:-"3.13"}
 RABBITMQ_IMAGE=${RABBITMQ_IMAGE:-"docker.io/rabbitmq"}
+VALKEY_VERSION=${VALKEY_VERSION:-"latest"}
```
Comment (hexfusion) [inline test/scripts/deploy_with_helm.sh:8]: should we consider pinning this to a stable version?

**Review by hexfusion — COMMENTED** [review]
Few small suggestions overall it feels sane. The only general question is can a user choose to use a Redis server with full support with this valkey client?

**Review by hexfusion — APPROVED** [review]

**Review by hexfusion — APPROVED** [review]
